File: 977.Squares-of-a-Sorted-Array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function (nums) {
  const n = nums.length;
  const result = Array(n);
  let start = 0;
  let end = n - 1;
  let i = n - 1;

  while (i >= 0) {
    const startSqr = nums[start] ** 2;
    const endSqr = nums[end] ** 2;

    if (startSqr > endSqr) {
      result[i] = startSqr;
      start += 1;
    } else {
      result[i] = endSqr;
      end -= 1;
    }

    i -= 1;
  }

  return result;
};

File: python/977.Squares-of-a-Sorted-Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [0] * n
        i = n - 1
        start = 0
        end = n - 1

        while start <= end:
            sSqr = nums[start] ** 2
            eSqr = nums[end] ** 2

            if sSqr > eSqr:
                result[i] = sSqr
                start += 1
                i -= 1
            else:
                result[i] = eSqr
                end -= 1
                i -= 1
        
        return result